plugin/evm/validators_normalizer_test.go

```diff
@@ -0,0 +1,173 @@
+package evm
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/flare-foundation/flare/ids"
+	"github.com/flare-foundation/flare/utils/logging"
+)
+
+type TestValidatorsNormalizer struct {
+	ByEpochFunc         func(uint64) (map[ids.ShortID]uint64, error)
+	CalcWeightRatioFunc func(map[ids.ShortID]uint64) map[ids.ShortID]uint64
+}
+
+func (t TestValidatorsNormalizer) ByEpoch(e uint64) (map[ids.ShortID]uint64, error) {
+	return t.ByEpochFunc(e)
+}
+
+func (t TestValidatorsNormalizer) calcWeightRatio(vals map[ids.ShortID]uint64) map[ids.ShortID]uint64 {
+	return t.CalcWeightRatioFunc(vals)
+}
+
+func TestValidatorsNormalizer_ByEpoch(t *testing.T) {
+	t.Run("nominal case", func(t *testing.T) {
+		var (
+			testValidators     = validatorsData
+			testValidatorsCopy = testValidators.duplicate()
+			calls              int
+		)
+
+		mock := TestValidatorsNormalizer{
+			ByEpochFunc: func(e uint64) (map[ids.ShortID]uint64, error) {
+				return testValidators[e], nil
+			},
+			CalcWeightRatioFunc: func(vals map[ids.ShortID]uint64) map[ids.ShortID]uint64 {
+				for i, _ := range vals {
+					vals[i] = uint64(0)
+				}
+				calls++
+				return vals
+			},
+		}
+
+		valCache := NewValidatorsCache(mock, WithCacheSize(128))
+		for k, v := range testValidators {
+			valCache.cache.Add(k, v)
+		}
+		valNormalizer := NewValidatorsNormalizer(&logging.Log{}, valCache)
+
+		for i := 0; i < len(validatorsData); i++ {
+			expected := mock.CalcWeightRatioFunc(testValidatorsCopy[uint64(i)])
+			result, err := valNormalizer.ByEpoch(uint64(i))
+			require.NoError(t, err)
+
+			assert.ElementsMatch(t, expected, result)
+		}
+		assert.Equal(t, 2, calls)
+	})
```

Comment:
I think there's still one thing that I haven't managed to explain properly :p Maybe we can have another call if it's unclear, but here is how I would have implemented this test:

```suggestion
	t.Run("nominal case", func(t *testing.T) {
		testValidators := validatorsData
		testWeightRatios := map[ids.ShortID]uint64{
			ids.ShortID{1}: 13,
			ids.ShortID{2}: 37,
		}
		
		var calls int
		mock := TestValidatorsNormalizer{
			ByEpochFunc: func(e uint64) (map[ids.ShortID]uint64, error) {
				return testValidators[e], nil
			},
			CalcWeightRatioFunc: func(vals map[ids.ShortID]uint64) map[ids.ShortID]uint64 {
				return testWeightRatios
			},
		}

		valCache := NewValidatorsCache(mock, WithCacheSize(128))
		for k, v := range testValidators {
			valCache.cache.Add(k, v)
		}
		valNormalizer := NewValidatorsNormalizer(&logging.Log{}, valCache)

		for i := 0; i < len(validatorsData); i++ {
			got, err := valNormalizer.ByEpoch(uint64(i))
			require.NoError(t, err)

			assert.ElementsMatch(t, testWeightRatios, got)
		}
		assert.Equal(t, 2, calls)
	})
```

* We no longer rely on the mock to give us the expected value, it's simply hardcoded
* We no longer modify the map we receive in the mock
* We no longer need the `duplicate` helper to duplicate the validator map
* Also some small style changes which are not very important
	* renamed result to got for consistency with other tests
	* removed `var` block for clarity

WDYT? If this seems good to you, feel free to apply those suggestions to other tests as well ^^

Reply:
Corrected here https://github.com/flare-foundation/coreth/pull/30/commits/92d487996acec507b6bb402cb312a45b734f4ec9
Thanks very much for the explanation, makes sense!

Reply:
As I said though, if this is good for you, those suggestions should also be applied to the other subtests of this function, and then the `duplicate` helper can even be removed entirely 👍

Reply:
Yes, I know.
Added them here https://github.com/flare-foundation/coreth/pull/30/commits/900a081da60254efa28e6b184e37f1dd40a1d54f
Thanks!